diag_manager/fms_diag_time_reduction.F90

```diff
@@ -0,0 +1,218 @@
+!* FMS is free software: you can redistribute it and/or modify it under
+!* the terms of the GNU Lesser General Public License as published by
+!* the Free Software Foundation, either version 3 of the License, or (at
+!* your option) any later version.
+!*
+!* FMS is distributed in the hope that it will be useful, but WITHOUT
+!* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+!* FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+!* for more details.
+!*
+!* You should have received a copy of the GNU Lesser General Public
+!* License along with FMS.  If not, see <http://www.gnu.org/licenses/>.
+!***********************************************************************
+
+!> @defgroup fms_diag_time_reduction_mod fms_diag_time_reduction_mod
+!> @ingroup diag_manager
+!> @brief fms_diag_time_reduction_mod defines classes encapsulating the diag_manager
+!! time redution types.
+!!
+!> @author Miguel Zuniga
+!!
+!> @file
+!> @brief File for @ref fms_diag_time_reduction_mod
+!> @addtogroup fms_diag_time_reduction_mod
+!> @{
+MODULE fms_diag_time_reduction_mod
+
+  USE diag_data_mod, only: EVERY_TIME
+   !!use diag_data_mod, only: time_min, time_max, time_sum, time_rms, time_average, time_none, time_power, &
+   !!& time_diurnal, every_time
+  USE fms_mod, ONLY: error_mesg, FATAL
+
+   implicit none
+
+   !!These parametes are the possible kinds of time reduction operations.
+   !!Note that sometimes one kind inplies another.
+   !!TODO: should they be put in diag_data ?
+   !!TODO:
+   !!TODO: time_diurnal "not really" same kind as others, so remove?
+   INTEGER, PARAMETER :: time_none    = 0 !< There is no reduction method
+   INTEGER, PARAMETER :: time_average = 1 !< The reduction method is avera
+   INTEGER, PARAMETER :: time_rms     = 2 !< The reduction method is rms
+   INTEGER, PARAMETER :: time_max     = 3 !< The reduction method is max
+   INTEGER, PARAMETER :: time_min     = 4 !< The reduction method is min
+   INTEGER, PARAMETER :: time_sum     = 5 !< The reudction method is sum
+   INTEGER, PARAMETER :: time_diurnal = 6 !< The reduction method is diurnal
+   INTEGER, PARAMETER :: time_power   = 7 !< The reduction method is power
+
+!> @brief Class time_reduction_type has an encapsulation of the "Fortran enum" time
+!! reduction integer parameters, plus an encapsulation of the groupings of
+!! the time reduction types. It is inteded to provide some of the functionality
+!! that was coded in the legacy function diag_data.F90:init_output_fields.
+!! The functionality in the end is used by send_data in (EFFICIENT) do loops calling
+!! the weighting or math functions to update buffers.
+!!  the The integer parameters above are the legal time_reduction_types,
+!! but they are not necessarily mutually exclusive in some contexts.
+!!
+!> @addtogroup fms_diag_time_reduction_mod
+   TYPE time_reduction_type
+      integer , private :: the_type !< The time reduction type; integer as per diag_data_mod entries.
```

Comment:
This is an `integer`, not a `type`.  This confused me while I was reading the code.  Maybe `time_reduction` or `time_reduction_op` would be a better name.  Also, the values are in this module, not diag_data like the comment implies.

Reply:
I changed "the_type" to "the_time_reduction". Modified the comment.